README
Wx-TreeListCtrl
===============


A simple wrapper around the wxWidgets class wxTreeListCtrl.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires wxWidgets and wxPerl:

COPYRIGHT AND LICENCE

Copyright (C) 2006 - 2010 by Mark Wardell

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.

This package includes the wxTreeListCtrl source code which you may use
according to the the wxWidgets license.

The authors of the wxTreeListCtrl package are:
	Robert Roebling,
	Julian Smart,
	Alberto Griggio,
	Vadim Zeitlin,
	Otto Wyss,
	Guru Kathiresan



Changes
Revision history for Perl extension Wx::TreeListCtrl.

0.08  Feb  10  2010
    - namespace fix
    
0.07  Feb  10  2010
    - Fixed possible Wx::TreeListColumnInfo thread memory leak
    - Added some tests.   

0.06  Feb   2  2010
      A patched non official version.
      get the official version from search.cpan.org
    - Updated to use latest source from wxTreeListCtrl
      in wxPython dist but added wxTR_SHOW_ROOT_LABEL_ONLY
      back in.
    - added GetColumn SetColumn GetCurrentItem SetCurrentItem
    - confirmed compilation on wxMAC (carbon), wxMSW, wxGTK (linux)

0.051 A patched non official version.
      get the official version from search.cpan.org

0.05  Mon Aug  21 2006
    - changed build system in accordance with Wx 0.57
	- removed script/pipe.pl

0.04  Fri Aug  18 2006
    - added lib/Wx/DemoModules/wxTreeList.pm

0.03  Thu Aug  17 2006
    - added EXPORT_TAG 'treelist'

0.02  Thu Aug  17 2006
    - fixed exporting of constants

0.01  Tue Aug  8 10:28:54 2006
	- first attempt at wrapping wxTreeListCtrl



t/Tests_Helper.pm
#############################################################################
#  File from the 0.97 version Wx distribution
#
#############################################################################
## Name:        t/Tests_Helper.pm
## Purpose:     some test helper functions
## Author:      Mattia Barbon
## Modified by:
## Created:     02/06/2001
## SVN-ID:      $Id: Tests_Helper.pm 22 2010-02-10 05:09:05Z  $
## Copyright:   (c) 2001-2003, 2005 Mattia Barbon
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################

package Tests_Helper;

use strict;
use Wx;
require Exporter;

use Test::More ();
*ok = \&Test::More::ok;
*is = \&Test::More::is;
*diag = \&Test::More::diag;

use vars qw(@ISA %EXPORT_TAGS @EXPORT_OK);

@ISA = qw(Exporter);

%EXPORT_TAGS =
  ( inheritance => [ qw(test_inheritance test_inheritance_all
                        test_inheritance_start test_inheritance_end) ],
  );

@EXPORT_OK = ( qw(test_app app_timeout test_frame in_frame),
               @{$EXPORT_TAGS{inheritance}} );

sub in_frame($) {
  my $callback = shift;
  my $sub = sub {
    my $frame = Tests_Helper_Frame->new( $callback );

    $frame->Show( 1 );
  };

  test_app( $sub );

  Wx::wxTheApp->MainLoop;
}

sub app_timeout($) {
  test_app( sub {
              my $frame = Wx::Frame->new( undef, -1, 'test' );
              my $timer = Wx::Timer->new( $frame );

              Wx::Event::EVT_TIMER( $frame, -1, sub {
                                      Wx::wxTheApp()->ExitMainLoop;
                                      $frame->Destroy;
                                    } );

              $timer->Start( 500, 1 );
              Wx::WakeUpIdle();
              $frame->Show( 1 );
            } );
}

sub test_app {
  my $function = shift;

  return Tests_Helper_App->new( $function );
}

sub test_frame {
  my $class = shift;
  my @params = @_;

  my $function = sub {
    my $frame = $class->new( @params );
  };

  my $app = Tests_Helper_App->new( $function );

  return $app;
}

sub test_inheritance {
  my( %perl_inheritance, %cpp_inheritance );

 LOOP: foreach my $i ( @_ ) {
    my $key = $i;
    my $cn = "wx${key}";
    my $ci = Wx::ClassInfo::FindClass( $cn ) or next LOOP;

    while ( 1 ) {
      push @{$cpp_inheritance{$key}}, cpp_2_perl( $cn );

      last unless $ci;
      $cn = $ci->GetBaseClassName1();
      last unless $cn;
      $ci = Wx::ClassInfo::FindClass( $cn );
    }

    my $class = $key;

    while ( $class ) {
      push @{$perl_inheritance{$key}}, "Wx::$class";

      last unless exists $Wx::{"${class}::"}{ISA} && 
        @{ $Wx::{"${class}::"}{ISA} };
      die $class unless defined @{ $Wx::{"${class}::"}{ISA} }[0];
      $class = substr @{ $Wx::{"${class}::"}{ISA} }[0], 4;
    }
  }

 CLASSES: foreach my $i ( keys %perl_inheritance ) {
    my $pi = $perl_inheritance{$i};
    my $ci = $cpp_inheritance{$i};
    my @pi = @$pi;
    my @ci =  @$ci;

  COMPARE: while ( @ci ) {
      my( $c_class ) = shift @ci;
      next if $c_class =~ m/Wx::Generic(?:ListCtrl|ImageList)/;
      next if $c_class =~ m/(?:Base|GTK|X11)$/;
      next if $c_class =~ m/StatusBar/; #FIXME// ad hoc
      next if $c_class eq 'Wx::Object';
      my( $p_class );

      while ( @pi ) {
        $p_class = shift @pi;
        next COMPARE if $c_class eq $p_class;
      }

      ok( 0, $pi->[0] . ' inheritance chain' );
      diag( "C++ : @{$ci}" );
      diag( "Perl: @{$pi}" );

      next CLASSES;
    }

    ok( 1,  $pi->[0] . ' inheritance chain' );
  }
}

{
  my %classes_skip;

  sub test_inheritance_start {
    foreach my $i ( keys %Wx:: ) {
      next unless $i =~ m/^([^_].*)::$/;
      $classes_skip{$1} = 1;
    }
  }

  sub test_inheritance_end {
    my @classes;

    foreach my $i ( keys %Wx:: ) {
      next unless $i =~ m/^([^_].*)::$/;
      next if exists $classes_skip{$1};
      push @classes, $1;
    }

    test_inheritance( @classes );
  }
}

sub test_inheritance_all {
  my @classes;

  foreach my $i ( keys %Wx:: ) {
    next unless $i =~ m/^([^_].*)::$/;
    push @classes, $1;
  }

  test_inheritance( @classes );
}

# utility

sub perl_2_cpp {
  my( $v ) = $_[0];

  $v =~ s/^Wx::/wx/;

  $v;
}

sub cpp_2_perl {
  my( $v ) = $_[0];

  $v =~ s/^wx/Wx::/;

  $v;
}

package Tests_Helper_App;

use base 'Wx::App';

my $on_init;

sub new {
  my $class = shift;
  my $function = shift;
  $on_init = $function;
  my $this = $class->SUPER::new( @_ );
  $this->SetExitOnFrameDelete(1);
  return $this;
}

sub OnInit {
  &$on_init;

  return 1;
}

package Tests_Helper_Frame;

use base 'Wx::Frame';

sub new {
  my $ref = shift;
  my $callback = shift;
  my $self = $ref->SUPER::new( undef, -1, "Test Frame" );
  my $timer = Wx::Timer->new( $self );

  Wx::Event::EVT_TIMER( $self, -1, sub {
                            &$callback( $self, $_[1] );
                            $self->Destroy;
                        } );

  $timer->Start( 500, 1 );
  Wx::WakeUpIdle();

  return $self;
}

sub Destroy {
    my $self = shift;

    $self->SUPER::Destroy;
    Wx::wxTheApp()->ExitMainLoop;
    Wx::WakeUpIdle();
}

1;

# Local variables: #
# mode: cperl #
# End: #


cpp/wxtreelist.h

class wxPliTreeListCtrl:public wxTreeListCtrl
{
    WXPLI_DECLARE_DYNAMIC_CLASS( wxPliTreeListCtrl );
    WXPLI_DECLARE_V_CBACK();
public:
    WXPLI_DEFAULT_CONSTRUCTOR( wxPliTreeListCtrl, "Wx::TreeListCtrl", true );
    WXPLI_CONSTRUCTOR_7( wxPliTreeListCtrl, "Wx::TreeListCtrl", true,
                         wxWindow*, wxWindowID, const wxPoint&,
                         const wxSize&, long, const wxValidator&,
                         const wxString& );

    virtual wxString OnGetItemText( wxTreeItemData* item, long column ) const;
    int OnCompareItems( const wxTreeItemId& item1, const wxTreeItemId& item2 );
    
};

wxString wxPliTreeListCtrl::OnGetItemText( wxTreeItemData* item, long column ) const
{
    dTHX;
    if( wxPliVirtualCallback_FindCallback( aTHX_ &m_callback, "OnGetItemText" ) )
    {
        SV* t1 = wxPli_non_object_2_sv( aTHX_ newSViv( 0 ), (void*)&item, "Wx::TreeItemData" );
        
        SV* ret = wxPliVirtualCallback_CallCallback( aTHX_ &m_callback, G_SCALAR,
                                                     "Sl", t1, column );
        
        sv_setiv( SvRV( t1 ), 0 );
        wxString val;
        WXSTRING_INPUT( val, char*, ret );
        SvREFCNT_dec( ret );
        SvREFCNT_dec( t1 );
        
        return val;
    }

    return wxTreeListCtrl::OnGetItemText( item, column );
}

int wxPliTreeListCtrl::OnCompareItems( const wxTreeItemId& item1,
                                       const wxTreeItemId& item2 )
{
    dTHX;
    if( wxPliVirtualCallback_FindCallback( aTHX_ &m_callback, "OnCompareItems" ) )
    {
        SV* t1 = wxPli_non_object_2_sv( aTHX_ newSViv( 0 ),
                                        (void*)&item1, "Wx::TreeItemId" );
        SV* t2 = wxPli_non_object_2_sv( aTHX_ newSViv( 0 ),
                                        (void*)&item2, "Wx::TreeItemId" );
        SV* ret = wxPliVirtualCallback_CallCallback
            ( aTHX_ &m_callback, G_SCALAR, "SS", t1, t2 );

        sv_setiv( SvRV( t1 ), 0 );
        sv_setiv( SvRV( t2 ), 0 );
        int val = SvIV( ret );
        SvREFCNT_dec( ret );
        SvREFCNT_dec( t1 );
        SvREFCNT_dec( t2 );

        return val;
    }
    else
        return wxTreeListCtrl::OnCompareItems( item1, item2 );
}

WXPLI_IMPLEMENT_DYNAMIC_CLASS( wxPliTreeListCtrl, wxTreeListCtrl );


cpp/tl_constants.cpp
/////////////////////////////////////////////////////////////////////////////
// Name:        tl_constants.cpp
// Purpose:     wxPerl constants
// Author:      Mark Wardell
// SVN ID:      $Id: tl_constants.cpp 22 2010-02-10 05:09:05Z  $
// Copyright:   (c) 2006 - 2010 Mark Wardell
// Licence:     This program is free software; you can redistribute it and/or
//              modify it under the same terms as Perl itself
/////////////////////////////////////////////////////////////////////////////


#include <cpp/constants.h>

double treelist_constant( const char* name, int arg )
{
    // !package: Wx
    // !parser: sub { $_[0] =~ m<^\s*r\w*\(\s*(\w+)\s*\);\s*(?://(.*))?$> }
    // !tag: grid
#define r( n ) \
    if( strEQ( name, #n ) ) \
        return n;

    WX_PL_CONSTANT_INIT();

    switch( fl )
    {
    case 'T':
        r( wxTL_MODE_NAV_FULLTREE );
        r( wxTL_MODE_NAV_EXPANDED );
        r( wxTL_MODE_NAV_VISIBLE );
        r( wxTL_MODE_NAV_LEVEL );
        r( wxTL_MODE_FIND_EXACT );
        r( wxTL_MODE_FIND_PARTIAL );
        r( wxTL_MODE_FIND_NOCASE );
        r( wxTR_HAS_BUTTONS );
        r( wxTR_NO_LINES );
        r( wxTR_LINES_AT_ROOT );
        r( wxTR_TWIST_BUTTONS );
        r( wxTR_MULTIPLE );
        r( wxTR_EXTENDED );
        r( wxTR_HAS_VARIABLE_ROW_HEIGHT );
        r( wxTR_EDIT_LABELS );
        r( wxTR_ROW_LINES );
        r( wxTR_HIDE_ROOT );
        r( wxTR_FULL_ROW_HIGHLIGHT );
        r( wxTR_DEFAULT_STYLE );
        r( wxTR_SINGLE );
        r( wxTR_NO_BUTTONS );
        r( wxTR_VIRTUAL );
        r( wxTR_COLUMN_LINES );
        r( wxTREE_HITTEST_ONITEMCOLUMN );
        r( wxTR_SHOW_ROOT_LABEL_ONLY );
        break;
    default:
        break;
    }
#undef r

  WX_PL_CONSTANT_CLEANUP();
}

wxPlConstants treelist_module( &treelist_constant );



lib/Wx/DemoModules/wxTreeListCtrl.pm
#############################################################################
## Name:        lib/Wx/DemoModules/wxTreeListCtrl.pm
## Purpose:     wxPerl demo helper for Wx::TreeListCtrl
## Author:      Mattia Barbon
## Modified by: Mark Dootson
## Created:     13/08/2006
## RCS-ID:      $Id: wxTreeListCtrl.pm 23 2010-02-10 05:17:34Z  $
## Copyright:   (c) 2005-2010 Mattia Barbon
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################


package Wx::DemoModules::wxTreeListCtrl;

use strict;
use base qw(Wx::Panel Class::Accessor::Fast);
use Wx::TreeListCtrl;
use Wx qw( :treelist :listctrl wxDefaultPosition wxDefaultSize wxVERTICAL wxNO_BORDER wxALL wxEXPAND);

our $VERSION = '0.08';

#use Wx::Event qw(EVT_SLIDER);

__PACKAGE__->mk_ro_accessors( qw(treelist) );

sub new {
    my( $class, $parent ) = @_;
    my $self = $class->SUPER::new( $parent );

    my $tree = $self->{treelist} = Wx::DemoModules::wxTreeListCtrl::Control->new( $self, -1,
        wxDefaultPosition, [400,200],
        wxTR_HIDE_ROOT | wxTR_ROW_LINES | wxTR_HAS_BUTTONS
        | wxTR_FULL_ROW_HIGHLIGHT | wxTR_SHOW_ROOT_LABEL_ONLY | wxTR_NO_LINES
    );


    # now add the columns
    if ($Wx::TreeListCtrl::VERSION > 0.06) {
        my $colinfo = Wx::TreeListColumnInfo->new("Column Three", 120, wxLIST_FORMAT_LEFT);
        $tree->AddColumn($colinfo);
        $colinfo->SetText("Column Two");
        $tree->InsertColumn(0, $colinfo);
        $colinfo->SetText("Column One");
        $tree->InsertColumn(0, $colinfo);
    } else {
        $tree->AddColumn( "Column Three",   120, wxLIST_FORMAT_LEFT );
        $tree->InsertColumn( 0, "Column Two",       120, wxLIST_FORMAT_LEFT );
        $tree->InsertColumn( 0, "Column One",       120, wxLIST_FORMAT_LEFT );
    }
    
    my $root = $tree->AddRoot( 'Root Item' );
    my $item1 = $tree->AppendItem( $root, 'First Top Level Tree Item Is Very Long' );
    # $tree->SetItemHeight( $item1, 120 );
    $tree->SetItemBold( $item1, 1 );
    $tree->SetItemTextColour( $item1, Wx::Colour->new( 22, 14, 135 ));
    $tree->SetItemBackgroundColour( $item1, Wx::Colour->new( 160, 184, 255 ));
    my $child1 = $tree->AppendItem( $item1, 'Child #1' );
    my $child2 = $tree->AppendItem( $item1, 'Child #2' );
    my $child3 = $tree->AppendItem( $item1, 'Child #3' );
        
        # call Column method directly
    $tree->SetItemColumnText( $child1, 1, "Child #1 - Column 2" );
    $tree->SetItemColumnText( $child1, 2, "Child #1 - Column 3" );
    $tree->SetItemColumnText( $child2, 1, "Child #2 - Column 2" );
        # call overloaded method - same result as above
    $tree->SetItemText( $child2, 2, "Child #2 - Column 3" );
    $tree->SetItemText( $child3, 1, "Child #3 - Column 2" );
    $tree->SetItemText( $child3, 2, "Child #3 - Column 3" );

    my $item2 = $tree->AppendItem( $root, 'Second Tree Item Is Also Long' );
    $tree->SetItemBold( $item2, 1 );
    $tree->SetItemTextColour( $item2, Wx::Colour->new( 178, 12, 48 ));
    $tree->SetItemBackgroundColour( $item2, Wx::Colour->new( 255, 211, 135 ));
    my $childA = $tree->AppendItem( $item2, 'Child A' );
    my $childB = $tree->AppendItem( $item2, 'Child B' );
    my $childC = $tree->AppendItem( $item2, 'Child C' );
    $tree->SetItemText( $childA, 1, "Child A - Column 2" );
    $tree->SetItemText( $childA, 2, "Child A - Column 3" );
    $tree->SetItemText( $childB, 1, "Child B - Column 2" );
    $tree->SetItemText( $childB, 2, "Child B - Column 3" );
    $tree->SetItemText( $childC, 1, "Child C - Column 2" );
    $tree->SetItemText( $childC, 2, "Child C - Column 3" );
        $tree->ExpandAll( $tree->GetRootItem );
        $tree->SortChildren($item1);
        $tree->SortChildren($item2);
        
    my $sizer = Wx::BoxSizer->new(wxVERTICAL);
    $sizer->Add($tree,1,wxALL|wxEXPAND, 5);
    $self->SetSizer($sizer);
    $self->Layout;
        
    return $self;
}

sub add_to_tags { qw(controls) }
sub title { 'wxTreeListCtrl' }

package Wx::DemoModules::wxTreeListCtrl::Control;
  
use strict;
use Wx::TreeListCtrl;
use base qw( Wx::TreeListCtrl );

sub new {
    my $class = shift;
    my $self = $class->SUPER::new( @_ );
    return $self;
}

sub OnCompareItems {
    my ($self, $item1, $item2) = @_;
    my $text1 = $self->GetItemText( $item1 );
    my $text2 = $self->GetItemText( $item2 );
    Wx::LogMessage("Wx::TreeListCtrl Compare Items; %s : %s", $text1, $text2);
    return $text1 cmp $text2;
}

1;



lib/Wx/TreeListCtrl.pm
#############################################################################
## Name:        TreeListCtrl.pm
## Purpose:     Wx::TreeListControl
## Author:      Mark Wardell
## Modified by: Mark Dootson
## RCS-ID:      $Id: TreeListCtrl.pm 22 2010-02-10 05:09:05Z  $
## Copyright:   (c) 2006-2010 Mark Wardell
## Licence:     This program is free software; you can redistribute it and/or
##              modify it under the same terms as Perl itself
#############################################################################

package Wx::TreeListCtrl;
use strict;
use warnings;
use Wx;

our $VERSION = '0.08';

our @constants = qw(
     wxTL_MODE_NAV_FULLTREE
     wxTL_MODE_NAV_EXPANDED
     wxTL_MODE_NAV_VISIBLE
     wxTL_MODE_NAV_LEVEL
     wxTL_MODE_FIND_EXACT
     wxTL_MODE_FIND_PARTIAL 
     wxTL_MODE_FIND_NOCASE
     wxTR_HAS_BUTTONS
     wxTR_NO_LINES
     wxTR_LINES_AT_ROOT
     wxTR_TWIST_BUTTONS
     wxTR_MULTIPLE
     wxTR_EXTENDED
     wxTR_HAS_VARIABLE_ROW_HEIGHT
     wxTR_EDIT_LABELS
     wxTR_ROW_LINES
     wxTR_HIDE_ROOT
     wxTR_FULL_ROW_HIGHLIGHT
     wxTR_DEFAULT_STYLE
     wxTR_SINGLE
     wxTR_NO_BUTTONS
     wxTR_VIRTUAL
     wxTR_COLUMN_LINES
     wxTREE_HITTEST_ONITEMCOLUMN
     wxTR_SHOW_ROOT_LABEL_ONLY
);

push @Wx::EXPORT_OK, @constants;
$Wx::EXPORT_TAGS{'treelist'} = [ @constants ];

Wx::wx_boot( 'Wx::TreeListCtrl', $VERSION );

#
# confirm inheritance tree
#

no strict;

package Wx::TreeListCtrl;  @ISA = qw( Wx::Control );

1;
__END__

=head1 NAME

Wx::TreeListCtrl - interface to the Wx::TreeListCtrl class

=head1 VERSION 0.08

=head1 SYNOPSIS

  use Wx::TreeListCtrl;

=head1 DESCRIPTION

Wx::TreeListCtrl is a wrapper for the wxTreeListCtrl class in the wxWidgets GUI toolkit.

=head2 EXPORT

None by default.

=head2 Exportable constants

  wxTL_MODE_NAV_FULLTREE
  wxTL_MODE_NAV_EXPANDED
  wxTL_MODE_NAV_VISIBLE
  wxTL_MODE_NAV_LEVEL
  wxTL_MODE_FIND_EXACT
  wxTL_MODE_FIND_PARTIAL 
  wxTL_MODE_FIND_NOCASE
  wxTR_HAS_BUTTONS
  wxTR_NO_LINES
  wxTR_LINES_AT_ROOT
  wxTR_TWIST_BUTTONS
  wxTR_MULTIPLE
  wxTR_EXTENDED
  wxTR_HAS_VARIABLE_ROW_HEIGHT
  wxTR_EDIT_LABELS
  wxTR_ROW_LINES
  wxTR_HIDE_ROOT
  wxTR_FULL_ROW_HIGHLIGHT
  wxTR_DEFAULT_STYLE
  wxTR_SINGLE
  wxTR_NO_BUTTONS
  wxTR_VIRTUAL
  wxTR_COLUMN_LINES
  wxTREE_HITTEST_ONITEMCOLUMN
  wxTR_SHOW_ROOT_LABEL_ONLY  

=head1 DOCUMENTATION

  The main Wx::TreeListCtrl has the same interface as the wxPython module for
  wxTreeListCtrl which is available here:
    
L<http://wxcode.sourceforge.net/components/treelistctrl/reference.html>
  
  Wx::TreeListColumnInfo is also available
  
  my $info = Wx::TreeListColumnInfo->new( coltext, width, flags, imageindex, shown, editable);
  
     only coltext is required, defaults for other items:
     width      = 100
     flags      = wxALIGN_LEFT
     imageindex = -1
     shown      = 1 (true)
     editable   = 0 (false)
  
  usage:
  
  $treelist->AddColumn($info);
  $treelist->InsertColumn(2, $info);
  $treelist->SetColumn(3, $info);
  
  # set methods have equivalent get methods
  my $info = $treelist->GetColumn(4); 
  $info->SetText('Column Four');
  $info->SetWidth('50');
  $info->SetAlignment(wxALIGN_RIGHT);
  $info->SetImage(1);
  $info->SetSelectedImage(2);
  $info->SetShown(1);
  $info->SetEditable(0);
  $treelist->SetColumn(4, $info);
  $info->SetText('Column Five');
  $treelist->SetColumn(5, $info);
  
  # note - changing a TreeListColumnInfo object will not
  # affect the underlying object. You must 'SetColumn' to
  # change the underlying data.
  
  # the two boolean members use 'Is' as a get method
  $info->IsShown
  $info->IsEditable
  


=head1 AUTHOR

Mark Wardell <mwardell@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 - 2010 by Mark Wardell

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.

This package includes the wxTreeListCtrl source code which you may use
according to the the wxWidgets license.

The authors of the wxTreeListCtrl package are:
    Robert Roebling,
    Julian Smart,
    Alberto Griggio,
    Vadim Zeitlin,
    Otto Wyss,
    Guru Kathiresan

=cut
